README
Copyright 2010 - James Townley & Brian Richardson
Release under Apache-BSD style license

Version: 0.0.1

----------------------------------
Description
----------------------------------

Wrapper for PyMongo Connection that offers additional application level monitoring.
When connecting to multiple MongoDB servers, allows you to verify connectivity to
MongoDB slave servers.

----------------------------------
Problem this solves
----------------------------------

Although the built in monitoring page for mongo is awesome it misses one situation.
This is when the mongo servers can talk to each other and the app can talk to the
master but the application has not been verified to be able to communicate with
the slave.

         _______________
        |               |
        |  APPLICATION  |
        |_______________|
           /         \
          /          (X)
  _______/___       ___\_______
 |           |     |           |
 |  Master   |-----|   Slave   |
 |___________|     |___________|


----------------------------------
Usage
----------------------------------

Use it as you would pymongo and call the check health method on connection

    import pymongo_frisk as pymongo

    connection = pymongo.connection.Connection.from_uri("mongo://username:password@host1,host2/database")
    results = connection.check_health()

results is a dictionary as follows:
{'db_master_url': 'host2', 'db_slave_url': 'host1', 'db_master_can_write': True, 'db_slave_can_read': True, 'db_master_can_read': True}




RELEASE_NOTES
Version 0.0.1 - Released 2010-07-16
- Initial Release

Version 0.0.2 - Released 2010-10-25
- Fixed : Used single field in collection instead of dropping and re-adding the collection each time. This was to counteract a problem with MongoDB crashing after many drop and re-adds.

Version 0.0.4 - Release 2011-01-10
- Updated to work with Replica Sets. Added the new class FriskConnection which extends pymongo Connection with our check_health() method

Version 0.0.5 - Release 2011-01-19
- Fixed: Failed to return if slave is "Write logged" due to bug in Mongo see https://jira.mongodb.org/browse/SERVER-2278
- New Feature: Write Locked slaves will now be listed as Write Locked in status


src/tests.py
import unittest, datetime
from pymongo_frisk import PyMongoFrisk as PMF, FriskConnection
from mock import patch, Mock
import pymongo.errors

class FriskConnectionStub(FriskConnection):
    __port = 27017
    __host = 'localhost'
    __nodes = set([('localhost', 27017), ('localhost', 27018), ('localhost', 27019)])
    
    def __init__(self, host=None, port=None, pool_size=None,
             auto_start_request=None, timeout=None, slave_okay=False,
             network_timeout=None, document_class=dict, tz_aware=False,
             _connect=True):
        self.database = DatabaseStub()
  
    @property
    def nodes(self):
        return self.__nodes
    
    @property
    def port(self):
        return self.__port
        
    @property
    def host(self):
        return self.__host
    
    def _get_datetime_now_microseconds(self):
        return 1
    
    def _get_new_uuid(self):
        return 1
        
class FriskConnectionTest(unittest.TestCase):

    host1 = 'localhost:27017'
    host2 = 'localhost:27018'
    host3 = 'localhost:27019'
    hosts = [host1,host2,host3] 
    health = {}
    connection = None
      
    def setUp(self):
        self.connection = FriskConnectionStub(self.hosts)
        
    @patch.object(pymongo.collection.Collection, 'remove')
    @patch.object(pymongo.collection.Collection, 'save')
    def test_check_health_shouldReturnMasterHost_whenConnectionCreated(self, mock_save, mock_remove):
        self.health =  self.connection.check_health()      
        self.assertEquals(self.host1, self.health['db_master_host'])      

    @patch.object(pymongo.collection.Collection, 'find_one')
    @patch.object(pymongo.collection.Collection, 'remove')
    @patch.object(pymongo.collection.Collection, 'save')
    def test_check_health_shouldReportCanWriteToMaster_whenMasterIsUp(self, mock_save, mock_remove, mock_find_one):
        mock_find_one.return_value={"_id":1, 'date': 1}
        self.health =  self.connection.check_health()
        self.assertEquals(True, self.health['db_master_can_write'])

    @patch.object(pymongo.collection.Collection, 'find_one')
    @patch.object(pymongo.collection.Collection, 'remove')
    @patch.object(pymongo.collection.Collection, 'save')
    def test_check_health_shouldReportCanWriteToMasterFailed_whenReadFromMasterDoesNotMatchDataWritten(self, mock_save, mock_remove, mock_find_one):
        mock_find_one.return_value={"_id":1, 'date': 1234}
        self.health =  self.connection.check_health()
        self.assertEquals(False, self.health['db_master_can_write'])

    @patch.object(pymongo.database.Database, 'collection_names')
    @patch.object(pymongo.collection.Collection, 'remove')
    @patch.object(pymongo.collection.Collection, 'save')    
    def test_check_health_shouldReportCanReadFromMaster_whenMasterIsUp(self, mock_save, mock_remove, mock_database_collection_names):
        mock_database_collection_names.return_value=['collection1','collection2']
        self.health =  self.connection.check_health()
        self.assertEquals(True, self.health['db_master_can_read'])

    @patch.object(pymongo.database.Database, 'collection_names')
    @patch.object(pymongo.collection.Collection, 'remove')
    @patch.object(pymongo.collection.Collection, 'save')    
    def test_check_health_shouldReportCanReadFromMasterFailed_whenReadFromMasterReturnsEmptyCollectionList(self, mock_save, mock_remove, mock_database_collection_names):
        mock_database_collection_names.return_value=[]
        self.health =  self.connection.check_health()
        self.assertEquals(False, self.health['db_master_can_read'])

    @patch.object(pymongo.collection.Collection, 'remove')
    @patch.object(pymongo.collection.Collection, 'save')            
    def test_check_health_shouldReturnListOfSlaveHosts_whenMultipleNodesInReplicaSet(self, mock_save, mock_remove):
        self.health =  self.connection.check_health()
        self.assertTrue(self.host1 not in self.health['db_slave_hosts'])
        self.assertTrue(self.host2 in self.health['db_slave_hosts'])
        self.assertTrue(self.host3 in self.health['db_slave_hosts'])

    @patch('pymongo.connection.Connection')
    @patch.object(pymongo.collection.Collection, 'find_one')
    @patch.object(pymongo.collection.Collection, 'remove')
    @patch.object(pymongo.collection.Collection, 'save')                
    def test_check_health_shouldReportCanReadFromAllSlaves_whenAllSlavesAreUp(self, mock_save, mock_remove, mock_find_one, mock_connection):
        mock_find_one.return_value={}
        mc = mock_connection.return_value
        mc.database_names.return_value = ['admin','local']
        class fakedCollection(object):
            def find_one(self):
                return {}
        fakedCollectionInstance = fakedCollection()
        mc.admin = {"$cmd.sys.inprog":fakedCollectionInstance}
        self.health =  self.connection.check_health()
        self.assertTrue((self.host2,True) in self.health['db_slaves_can_read'])
        self.assertTrue((self.host3,True) in self.health['db_slaves_can_read'])

    @patch('pymongo.connection.Connection')
    @patch.object(pymongo.collection.Collection, 'find_one')
    @patch.object(pymongo.collection.Collection, 'remove')
    @patch.object(pymongo.collection.Collection, 'save')
    def test_check_health_shouldReportCanReadFromAllSlaves_whenAllSlavesAreUpAndAreWriteLocked(self, mock_save, mock_remove, mock_find_one, mock_connection):
        mock_find_one.return_value={}
        mc = mock_connection.return_value
        mc.database_names.return_value = ['admin','local']
        class fakedCollection(object):
            def find_one(self):
                return {u"fsyncLock":1}
        fakedCollectionInstance = fakedCollection()
        mc.admin = {"$cmd.sys.inprog":fakedCollectionInstance}
        self.health =  self.connection.check_health()
        self.assertTrue((self.host2,"Write Locked") in self.health['db_slaves_can_read'])
        self.assertTrue((self.host3,"Write Locked") in self.health['db_slaves_can_read'])
 
    @patch('pymongo.connection.Connection')
    @patch.object(pymongo.collection.Collection, 'find_one')
    @patch.object(pymongo.collection.Collection, 'remove')
    @patch.object(pymongo.collection.Collection, 'save')                
    def test_check_health_shouldReportCanReadFromAllSlavesFailed_whenFailToQuerySlaves(self, mock_save, mock_remove, mock_find_one, mock_connection):
        mock_find_one.return_value={}
        mc = mock_connection.return_value
        mc.database_names.return_value = []
        self.health =  self.connection.check_health()
        self.assertTrue((self.host2,False) in self.health['db_slaves_can_read'])
        self.assertTrue((self.host3,False) in self.health['db_slaves_can_read'])    
        
class PyMongoFriskTest(unittest.TestCase):

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_connection_from_uri_calls_parent_with_same(self, mock_from_uri):
        test_uri = "mongo://user:password@url.com/database"
        mock_from_uri.return_value="Connection"
        pmf = PMF.from_uri(test_uri)
        mock_from_uri.assert_called_with(test_uri)

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_connection_from_uri_creates_expected_details(self, mock_from_uri):
        test_uri = "mongo://username:password@host1,host2/database"
        mock_from_uri.return_value="Connection"
        expected_connection = PMF.from_uri(test_uri)
        self.assertEquals('username',expected_connection._username)
        self.assertEquals('password',expected_connection._password)
        self.assertEquals('database',expected_connection._database)
        self.assertEquals(['host1','host2'],expected_connection._hosts)
        self.assertEquals('Connection', expected_connection._connection)

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_connection_from_uri_creates_expected_details_even_if_port_specified(self, mock_from_uri):
        test_uri = "mongo://username:password@host1:123,host2:456/database"
        mock_from_uri.return_value="Connection"
        expected_connection = PMF.from_uri(test_uri)
        self.assertEquals('username',expected_connection._username)
        self.assertEquals('password',expected_connection._password)
        self.assertEquals('database',expected_connection._database)
        self.assertEquals(['host1:123','host2:456'],expected_connection._hosts)
        self.assertEquals('Connection', expected_connection._connection)

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_connection_from_uri_creates_expected_details_given_one_host(self, mock_from_uri):
        test_uri = "mongo://username:password@host1/database"
        mock_from_uri.return_value="Connection"
        expected_connection = PMF.from_uri(test_uri)
        self.assertEquals('username',expected_connection._username)
        self.assertEquals('password',expected_connection._password)
        self.assertEquals('database',expected_connection._database)
        self.assertEquals(['host1'],expected_connection._hosts)
        self.assertEquals('Connection', expected_connection._connection)

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_connection_from_uri_creates_expected_details_given_no_db(self, mock_from_uri):
        test_uri = "mongo://username:password@host1"
        mock_from_uri.return_value="Connection"
        expected_connection = PMF.from_uri(test_uri)
        self.assertEquals('username',expected_connection._username)
        self.assertEquals('password',expected_connection._password)
        self.assertEquals(None,expected_connection._database)
        self.assertEquals(['host1'],expected_connection._hosts)
        self.assertEquals('Connection', expected_connection._connection)
#
    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_connection_from_uri_creates_expected_details_given_no_auth(self, mock_from_uri):
        test_uri = "mongo://host1"
        mock_from_uri.return_value="Connection"
        expected_connection = PMF.from_uri(test_uri)
        self.assertEquals(None,expected_connection._username)
        self.assertEquals(None,expected_connection._password)
        self.assertEquals(None,expected_connection._database)
        self.assertEquals(['host1'],expected_connection._hosts)
        self.assertEquals('Connection', expected_connection._connection)

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_connection_from_uri_throws_exception_when_invalid_uri(self, mock_from_uri):
        test_uri = "mongo://username@host/database"
        mock_from_uri.return_value="Connection"
        self.assertRaises(pymongo.errors.InvalidURI, PMF.from_uri, test_uri)

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_connection_from_uri_throws_exception_when_to_many_hosts(self, mock_from_uri):
        test_uri = "mongo://username:password@host1,host2,host3/database"
        mock_from_uri.return_value="Connection"
        self.assertRaises(pymongo.errors.InvalidURI, PMF.from_uri, test_uri)

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_check_health_returns_master_and_slave_urls(self, mock_from_uri):
        test_uri = "mongo://username:password@host1,host2/database"
        connection_stub = ConnectionStub()
        mock_from_uri.return_value=connection_stub
        pmf = PMF.from_uri(test_uri)
        health = pmf.check_health()
        self.assertEquals('host1',health['db_master_url'])
        self.assertEquals('host2',health['db_slave_url'])

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_check_health_returns_master_read_and_write_success(self, mock_from_uri):
        test_uri = "mongo://username:password@host1,host2/database"
        connection_stub = ConnectionStub()
        mock_from_uri.return_value=connection_stub
        pmf = PMF.from_uri(test_uri)
        health = pmf.check_health()
        self.assertTrue(health['db_master_can_read'])
        self.assertTrue(health['db_master_can_write'])
        self.assertTrue(connection_stub.database.collection.remove_called)

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_check_health_returns_master_read_fail_when_master_cannot_be_read(self, mock_from_uri):
        test_uri = "mongo://username:password@host1,host2/database"
        connection_stub = ConnectionStub()
        connection_stub.database.collections = []
        mock_from_uri.return_value=connection_stub
        pmf = PMF.from_uri(test_uri)
        health = pmf.check_health()
        self.assertFalse(health['db_master_can_read'])

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_check_health_returns_master_write_fail_when_master_cannot_write(self, mock_from_uri):
        test_uri = "mongo://username:password@host1,host2/database"
        connection_stub = ConnectionStub()
        connection_stub.database.collection.override_data = True
        connection_stub.database.collection.data = None
        mock_from_uri.return_value=connection_stub
        pmf = PMF.from_uri(test_uri)
        health = pmf.check_health()
        self.assertFalse(health['db_master_can_write'])

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_check_health_returns_slave_read_success(self, mock_from_uri):
        test_uri = "mongo://username:password@host1,host2/database"
        expected_slave_uri = "mongo://username:password@host2/database"
        master_connection_stub = ConnectionStub()
        slave_connection_stub = ConnectionStub()
        slave_connection_stub.host = 'host2'
        mock_from_uri.return_value=master_connection_stub
        pmf = PMF.from_uri(test_uri)
        mock_from_uri.return_value=slave_connection_stub
        health = pmf.check_health()

        mock_from_uri.assert_called_with(expected_slave_uri, slave_okay=True)
        self.assertTrue(health['db_slave_can_read'])
        self.assertTrue(slave_connection_stub.disconnect_called)

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_check_health_handles_no_slave(self, mock_from_uri):
        test_uri = "mongo://username:password@host1/database"
        connection_stub = ConnectionStub()
        mock_from_uri.return_value=connection_stub
        pmf = PMF.from_uri(test_uri)
        health = pmf.check_health()
        self.assertFalse(health['db_slave_can_read'])
        self.assertEquals(None, health['db_slave_url'])

    @patch.object(pymongo.connection.Connection, 'from_uri')
    def test_check_health_returns_slave_read_fail_when_slave_cannot_read(self, mock_from_uri):
        test_uri = "mongo://username:password@host1,host2/database"
        expected_slave_uri = "mongo://username:password@host2/database"
        master_connection_stub = ConnectionStub()
        slave_connection_stub = ConnectionStub()
        slave_connection_stub.database.collections = []
        slave_connection_stub.host = 'host2'
        mock_from_uri.return_value=master_connection_stub
        pmf = PMF.from_uri(test_uri)
        mock_from_uri.return_value=slave_connection_stub
        health = pmf.check_health()

        mock_from_uri.assert_called_with(expected_slave_uri, slave_okay=True)
        self.assertFalse(health['db_slave_can_read'])
        self.assertTrue(slave_connection_stub.disconnect_called)

class ConnectionStub(object):
    def __init__(self):
        self.defaults()

    def defaults(self):
        self.host= "host1"
        self.database = DatabaseStub()
        self.disconnect_called = False

    def __getitem__(self,name):
        return self.database

    def disconnect(self):
        self.disconnect_called = True

class DatabaseStub(object):
    def __init__(self):
        self.defaults()

    def defaults(self):
        self.collection = CollectionStub()
        self.drop_collection_called = False
        self.disconnect_called = False
        self.collections = ['friskmonitoring']

    def collection_names(self):
        return self.collections

    def __getitem__(self,name):
        return self.collection

    def drop_collection(self, collection_name):
        self.drop_collection_called = True

class CollectionStub(object):
    def __init__(self):
        self.defaults()

    def defaults(self):
        self.override_data = False
        self.remove_called = False

    def save(self, data):
        if not self.override_data:
            self.data = data

    def find_one(self, query):
        return self.data

    def remove(self, query):
        self.remove_called = True

if __name__=='__main__':
    unittest.main()


    

src/pymongo_frisk.py
import datetime, copy, uuid
import pymongo
from pymongo.connection import Connection as mongo_con
from pymongo import Connection

class FriskConnection(pymongo.Connection):
    """Extension to the PyMongo Connection class that adds a check_health method for verifying
    connectivity to the Master as well as ALL Slaves in the Replica Set
    """
    def check_health(self):
        """Returns the health of ALL nodes in a Replica Set in dictionary format.
        {'db_master_host': 'host1:27017', 'db_slave_hosts': ['host2:27018','host3:27019'], 
        'db_master_can_write': True, 'db_master_can_read': True,
        'db_slaves_can_read': [('host2',True),('host3',True)] }
        """
        db = 'monitoring'
        db_master_can_write = self._can_write_to_master(db)
        db_master_can_read = self._can_read_from_master(db)

        slaves = self._get_slave_hosts()
        db_slaves_can_read = self._can_read_from_slaves(slaves)

        return {'db_master_host': self.host + ':' + str(self.port),
                'db_slave_hosts': slaves,
                'db_master_can_read':db_master_can_read,
                'db_master_can_write':db_master_can_write,
                'db_slaves_can_read':db_slaves_can_read
                }

    def _get_slave_hosts(self):
        master_host = self.host
        master_port = self.port
        slaves = []
        for node in self.nodes:
            slave_host = node[0]
            slave_port = node[1]
            if '.' in slave_host:
                slave_host = str(slave_host).split(".",1)[0]
            if (slave_host != master_host or slave_port != master_port and not slave_host+':'+str(slave_port) in slaves):
                slaves.append(slave_host+':'+str(slave_port))
        return slaves
    
    def _can_read_from_slaves(self, slaves):
        db_slaves_can_read = []
        for slave in slaves:
            is_healthy = False
            slave_connection = None
            try:
                slave_connection = pymongo.connection.Connection(slave, slave_okay=True)
                stats=slave_connection.admin["$cmd.sys.inprog"].find_one()
                if(u"fsyncLock" not in stats.keys()):
                    if 'admin' in slave_connection.database_names():
                        is_healthy = True
                else:
                    is_healthy="Write Locked"
            except Exception as e:
                pass
            finally:
                if slave_connection:
                    slave_connection.disconnect()
            db_slaves_can_read.append((slave, is_healthy))
        return db_slaves_can_read

    def _can_write_to_master(self, db):
        master_db = self.db
        db_master_can_write = False
        try:
            id = self._get_new_uuid()
            test_data = {"_id":id, 'date': self._get_datetime_now_microseconds()}
            master_db['friskmonitoring'].save(test_data)
            db_master_can_write = master_db['friskmonitoring'].find_one({'_id':id}) == test_data
        except:
            pass
        finally:
            master_db['friskmonitoring'].remove({'_id':id})
        return db_master_can_write

    def _can_read_from_master(self, db):
        master_db = self.db
        db_master_can_read = False
        try:
            db_master_can_read = master_db.collection_names() != []
        except:
            pass
        return db_master_can_read

    def _get_datetime_now_microseconds(self):
        return datetime.datetime.now().microsecond
        
    def _get_new_uuid(self):
        return str(uuid.uuid1())

class PyMongoFrisk(object):
    """Wrapper for pymongo Connection allowing additional check_health method for checking connectivity 
    of master and slave in a Replica Pair running in Auth mode
    """
    def __init__(self, uri, **kw):
        self._uri = copy.copy(uri)
        self._parse_uri(uri)
        self._connection = mongo_con.from_uri(uri, **kw)

    @classmethod
    def from_uri(cls, uri, **kw):
        return cls(uri, **kw)

    def __getattr__(self, attr_name):
        return getattr(self._connection, attr_name)

    def __getitem__(self,name):
        return self._connection[name]

    def check_health(self):
        """Returns the health of a both nodes in a Replica Pair connection in dictionary format.
        {'db_master_url': 'host2', 'db_slave_url': 'host1', 
        'db_master_can_write': True, 'db_master_can_read': True, 
        'db_slave_can_read': True}
        """
        master_connection = self._connection[self._database]
        master = self._connection.host
        id = str(uuid.uuid1())
        test_data = {"_id":id, 'date': datetime.datetime.now().microsecond}

        db_master_can_write = False
        try:
            master_connection['friskmonitoring'].save(test_data)
            db_master_can_write = master_connection['friskmonitoring'].find_one({'_id':id}) == test_data
        except:
            pass
        finally:
            master_connection['friskmonitoring'].remove({'_id':id})

        db_master_can_read = False
        try:
            db_master_can_read = master_connection.collection_names() != []
        except:
            pass

        db_slave_can_read = False
        hosts = copy.copy(self._hosts)
        if len(hosts) > 1:
            hosts.remove(master)
            slave = hosts[0]
            slave_connection = None
            try:
                slave_uri = copy.copy(self._uri)
                if "," in slave_uri:
                    slave_uri = slave_uri.replace(master, '').replace(',','')
                    slave_uri = slave_uri.replace(master, '').replace(',','')
                slave_connection = mongo_con.from_uri(slave_uri,slave_okay=True)
                db_slave_can_read = slave_connection[self._database].collection_names() != []
            except:
                pass
            finally:
                if slave_connection:
                    slave_connection.disconnect()
        else:
            slave = None

        return {'db_master_url': master,
                'db_slave_url': slave,
                'db_master_can_read':db_master_can_read,
                'db_master_can_write':db_master_can_write,
                'db_slave_can_read':db_slave_can_read
                }

    def _parse_uri(self, uri):
        try:
            rest  = uri.split("://")[1]
            if "@" in rest:
                auth, rest = rest.split("@")
                self._username , self._password = auth.split(":")
            else:
                self._username , self._password = None, None
            if "/" in rest:
                hosts, self._database = rest.split("/")
            else:
                self._database = None
                hosts = rest
            self._hosts = hosts.split(",")
            if len(self._hosts) > 2:
                raise pymongo.errors.InvalidURI("Too many hosts")
        except:
            raise pymongo.errors.InvalidURI("Uri not in expected format")

Connection = PyMongoFrisk
